src/timely-util/src/progress.rs

```diff
@@ -27,3 +30,392 @@ where
 {
     proptest::collection::vec(any::<T>(), 1..11).prop_map(Into::into)
 }
+
+/// A partially ordered timestamp that is partitioned by an arbitrary number of partitions
+/// identified by `P`. The construction allows for efficient representation frontiers with
+/// Antichains.
+///
+/// ## The problem of partitioned timestamps
+///
+/// To understand this timestamp it helps to first think about how one would represent a
+/// partitioned, partially ordered timestamp type where each partition follows its own independent
+/// timeline in a natural way. Timely requires that all timestamps have a minimum element that is
+/// less than all other elements so you could express a partitioned timestamp like so:
+///
+/// ```rust
+/// enum Partitioned<P, T> {
+///     Bottom,
+///     Partition(P, T),
+/// }
+/// ```
+/// Laying out the order in a graph we'd see something like this:
+///
+/// ```text
+///           -----(P1, T0)---(P1, T1)--- ... ---(P1, Tn)
+///          / ----(P2, T0)---(P2, T1)--- ... ---(P2, Tn)
+///         / / ,--(P3, T0)---(P3, T1)--- ... ---(P3, Tn)
+/// Bottom-+-+-+     .
+///         \        .
+///          \       .
+///           '----(Pn, T0)---(Pn, T1)--- ... ---(Pn, Tn)
+/// ```
+///
+/// This timestamp has the problem that if you want to downgrade your operator's capability to
+/// indicate progress in one of the partitions, which implies dropping your `Bottom` capability,
+/// you are forced to also mint one capability per partition that you intend to produce data for in
+/// the future. If those partitions are unknown or, even worse, infinite, this type brings you to a
+/// dead end.
+///
+/// ## How this type works
+///
+/// The idea of this `Parititoned` timestamp is to extend the idea of the `Bottom` element above
+/// into a `Range` element that is parameterized by a lower and upper bound and represents a
+/// *range* of partitions at some timestamp T. The represented range has exclusive bounds.
+///
+/// The minimum timestamp of this type is `Range(Bottom, Bottom, T::minimum(0)` which is less than
```

Comment:
Should this be `Range(Bottom, Top, T::minimum())`? It seems like otherwise the described range is just empty.